src/lib.rs

```diff
@@ -1768,7 +1795,15 @@ mod tests {
 
     #[test]
     fn unfill_whitespace() {
-        assert_eq!(unfill("foo   bar").0, "foo   bar");
+        assert_eq!(unfill("foo   bar", LineEnding::LF).0, "foo   bar");
+    }
+
+    #[test]
+    fn unfill_crlf() {
```

Comment:
If we do the auto-detection like I suggest, could you then also please add another test which uses `refill` to change the line endings from `\n` to `\r\n`?

That is, I believe this should work:

```rust
assert_eq(refill("foo\nbar\n", Options::new(5).line_ending(LineEndings::CRLF)),
          "foo\r\nbar\r\n");
```

This would show that we correctly use the line ending from the input option to `refill` to join the lines after we use `unfill` to get the text unwrapped.

Reply:
> That is, I believe this should work

Except that we do not modify trailing line endings, hence it'll be 
```
"foo\r\nbar\n"
```
On one hand it makes sense (the lib's focus is text wrapping, not CRLF conversion), on the other it is a bit unexpected.

Reply:
Hmm... can you remind me why I decided to not modify the trailing line endings? :smile: 

Perhaps it's because I made `unfill` so simple: it expects a single wrapped paragraph and so a string with `\n\n` doesn't really fit into that model. But now that I think (briefly) about it again, it seems to me that it would be equally consistent to collapse the line endings (if they exist). So `"foo\n\n\n"` becomes `"foo\n"` and `"foo"` stays `"foo"`. But I might be missing something?

Reply:
As a user of `unfill` I'd prefer it to do all the magic and do just what I want.

From that point of view collapsing new lines seems preferrable to current behavior.

If we implement multiple paragraphs handling (as I mentioned somewhere above), then this behavior would seem the most consistent to me.

1. `unfill` would process a paragraph as it is doing now
2. as soon as it detects an empty line, it would reset all the internal state and repeat the processing on the remainder of the input

Any empty lines (including trailing line breaks), would then be collapsed and replaced by a common auto-detected value.

Reply:
Btw, about `unfill` doing all the magic: I think detecting more prefixes than just common list markers would be very useful. Particularly, it'd be great if unfill understood that it operates on a comment. Just stumbled on this while editing this PR in _helix_ editor (tell me about eating my own dogfood :smile:  )

Reply:
> As a user of `unfill` I'd prefer it to do all the magic and do just what is the most reasonable (if I need any advanced tailored behavior I'd probably implement it myself).
> 
> From that point of view collapsing new lines seems preferrable to current behavior.

I think I agree: you're saying that `unfill` should not keep training newlines like it does now. If so, then I agree :-)

> 1. `unfill` would process a paragraph as it is doing now
> 2. as soon as it detects an empty line, it would reset all the internal state and repeat the processing on the remainder of the input
> 
> Any empty lines (including trailing line breaks), would then be collapsed and replaced by a common auto-detected value.

I guess the multi-paragraph logic of `unfill` would be:

* a single `\n` is seen as a continuation of the current paragraph. It is removed and the text from both lines go into the current line.
* two or more `\n` is seen as a paragraph separator. The newlines are added to the unfolded text followed by the current line.

So

```
this is
the first
paragraph


second paragraph
still
second

third paragraph
```

becomes 

```
this is the first paragraph


second paragraph still second

third paragraph
```

Here `\n` and `\r\n` should be treated the same: if the input has a trailing `\r\n` or a `\r\n\r\n` paragraph separator, then the unfolded string should use `\r\n` as well. It will also indicate that this is the line ending by putting it `LineEnding::CRLF` into the `line_ending` field.

The caller (`refill`) can then split on `options.line_ending.as_str()` to get the individual _paragraphs_. It can fold them using the _new_ line ending and thus we should end up with a line ending conversion.

Does that seem reasonable?

Reply:
My thinking is the same, I believe we are almost¹ on the same page regarding this. Thanks for putting it into words.

> The caller (`refill`) can then split on options.line_ending.as_str() to get the individual paragraphs.

¹ - If you meant to say `unfill` here, then I agree with you. Otherwise I think I may be missing something.